File: archive/p/pascal/josephus_problem.pas
program JosephusProblem;

{$mode objfpc}{$H+}

uses
   Classes,
   SysUtils;

procedure ShowUsage;
begin
   Writeln('Usage: please input the total number of people and number of people to skip.');
   Halt(1);
end;

function ParseInteger(const S: string; out Value: integer): boolean;
begin
   Result := TryStrToInt(Trim(S), Value);
end;

function Josephus(TotalPeople, SkipCount: integer): integer;
var
   CurrentCount, HighestPowerOf2: integer;
begin
   if (SkipCount < 1) or (TotalPeople < 1) then ShowUsage;

   // If skip = 1, that means that survivor is the last person
   if SkipCount = 1 then
   begin
      Result := TotalPeople;
      Exit;
   end;

   if SkipCount = 2 then
   begin
      HighestPowerOf2 := 1;
      while (HighestPowerOf2 shl 1) <= TotalPeople do
         HighestPowerOf2 := HighestPowerOf2 shl 1;

      Result := 2 * (TotalPeople - HighestPowerOf2) + 1;
      Exit;
   end;

   Result := 0;

   for CurrentCount := 2 to TotalPeople do
      Result := (Result + SkipCount) mod CurrentCount;

   Result := Result + 1;
end;

var
   TotalPeople, SkipCount: integer;
begin
   if ParamCount <> 2 then
      ShowUsage;

   if not ParseInteger(ParamStr(1), TotalPeople) then
      ShowUsage;
   if not ParseInteger(ParamStr(2), SkipCount) then
      ShowUsage;

   Writeln(Josephus(TotalPeople, SkipCount));
end.

File: archive/e/elixir/josephus-problem.ex
defmodule Josephus do
  @spec josephus(n :: integer(), k :: integer()) :: integer()
  def josephus(1, _), do: 1
  def josephus(n, k), do: Integer.mod(josephus(n - 1, k) + k - 1, n) + 1
end

with [n, k] <- System.argv(),
     {n, ""} <- Integer.parse(n),
     {k, ""} <- Integer.parse(k) do
  Josephus.josephus(n, k) |> IO.puts()
else
  _ -> IO.puts("Usage: please input the total number of people and number of people to skip.")
end
